src/controls.js

```diff
@@ -201,63 +201,308 @@ Crafty.extend({
      */
-
     touchDispatch: function (e) {
-        var type,
-            lastEvent = Crafty.lastEvent;
-
-        if (e.type === "touchstart") type = "mousedown";
-        else if (e.type === "touchmove") type = "mousemove";
-        else if (e.type === "touchend") type = "mouseup";
-        else if (e.type === "touchcancel") type = "mouseup";
-        else if (e.type === "touchleave") type = "mouseup";
-
-        if (e.touches && e.touches.length) {
-            first = e.touches[0];
-        } else if (e.changedTouches && e.changedTouches.length) {
-            first = e.changedTouches[0];
-        }
-
-        var simulatedEvent = document.createEvent("MouseEvent");
-        simulatedEvent.initMouseEvent(type, true, true, window, 1,
-            first.screenX,
-            first.screenY,
-            first.clientX,
-            first.clientY,
-            false, false, false, false, 0, e.relatedTarget
-        );
-
-        first.target.dispatchEvent(simulatedEvent);
-
-        // trigger click when it should be triggered
-        if (lastEvent !== null && lastEvent.type == 'mousedown' && type == 'mouseup') {
-            type = 'click';
-
-            simulatedEvent = document.createEvent("MouseEvent");
-            simulatedEvent.initMouseEvent(type, true, true, window, 1,
-                first.screenX,
-                first.screenY,
-                first.clientX,
-                first.clientY,
-                false, false, false, false, 0, e.relatedTarget
-            );
-            first.target.dispatchEvent(simulatedEvent);
-        }
+        if (!Crafty.touchObjs) return;
+        
+        if (this._touchHandler.multitouch)
+            switch (e.type) {
+                case "touchstart":
+                    this._touchHandler.handleStart(e);
+                    break;
+                case "touchmove":
+                    this._touchHandler.handleMove(e);
+                    break;
+                case "touchleave": // touchleave is treated as touchend
+                case "touchend":
+                    this._touchHandler.handleEnd(e);
+                    break;
+                case "touchcancel":
+                    this._touchHandler.handleCancel(e);
+                    break;
+            }
+        else
+            this._touchHandler.mimicMouse(e);
 
         //Don't prevent default actions if target node is input or textarea.
-        if (e.target && e.target.nodeName !== 'INPUT' && e.target.nodeName !== 'TEXTAREA') {
+        if (e.target && e.target.nodeName !== 'INPUT' && e.target.nodeName !== 'TEXTAREA')
             if (e.preventDefault) {
                 e.preventDefault();
             } else {
                 e.returnValue = false;
             }
+    },
+    
+    _touchHandler: {
+        fingers: [], // keeps track of touching fingers
+        multitouch: false,
+        
+        handleStart: function (e) {
+            var touches = this.getTouchesFromEvent(e);
```

Comment:
Found the bug.  `getTouchesFromEvent` default to returning `e.touches`.  This will include *every* touch event, not just the ones recently updated.  The bug happens as follows:
- Touch entity A; event triggers properly.
- Touch entity B.  `handleStart()` is called, and retrieves the list of all current touches.  It then triggers a "TouchStart" event for all of them.  Since this includes the original touch on entity A, the bug manifests.

Something similar will happen with the other event handlers.